libnettest2.hpp

```diff
@@ -0,0 +1,1805 @@
+    for (auto &pair : ctx.test_helpers) {
+      auto &key = pair.first;
+      auto &values = pair.second;
+      for (auto &epnt : values) {
+        measurement["test_helpers"][key] = nlohmann::json::object();
+        measurement["test_helpers"][key]["address"] = epnt.address;
+        if (epnt.type == endpoint_type_onion) {
+          measurement["test_helpers"][key]["type"] = "onion";
+        } else if (epnt.type == endpoint_type_https) {
+          measurement["test_helpers"][key]["type"] = "https";
+        } else if (epnt.type == endpoint_type_cloudfront) {
+          measurement["test_helpers"][key]["type"] = "cloudfront";
+          measurement["test_helpers"][key]["front"] = epnt.front;
+        } else {
+          // NOTHING
+        }
+      }
+    }
+  }
+  measurement["test_name"] = nettest_.name();
+  measurement["test_start_time"] = test_start_time;
+  measurement["test_version"] = nettest_.version();
+  nlohmann::json test_keys;
+  auto measurement_start = std::chrono::steady_clock::now();
+  // TODO(bassosimone): make sure we correctly pass downstream the probe_ip
+  // such that the consumer tests could use it to scrub the IP. Currently the
+  // nettest with this requirements is WebConnectivity.
+  auto rv = nettest_.run(settings_, ctx, inputs[i], &test_keys, info);
+  {
+    auto current_time = std::chrono::steady_clock::now();
+    std::chrono::duration<double> elapsed = current_time - measurement_start;
+    measurement["test_runtime"] = elapsed.count();
+  }
+  // We fill the resolver_ip after the measurement. Doing that before may allow
+  // the nettest to overwrite the resolver_ip field set by us.
+  measurement["test_keys"] = test_keys;
+  measurement["test_keys"]["resolver_ip"] = settings_.save_real_resolver_ip
+                                                ? ctx.resolver_ip
+                                                : "";
+  if (!rv) {
+    // TODO(bassosimone): we should standardize the errors we emit. We can
+    // probably emit something along the lines of library_error.
+    emit_ev("failure.measurement", {
+        {"failure", "generic_error"},
+        {"idx", i},
+    });
+  }
+  do {
+    std::string str;
+    try {
+      str = measurement.dump();
+    } catch (const std::exception &e) {
+      LIBNETTEST2_EMIT_WARNING("run: cannot serialize JSON: " << e.what());
+      // TODO(bassosimone): This is MK passing us an invalid JSON. Should we
+      // submit something nonetheless as a form of telemetry? This is something
+      // I should probably discuss with @hellais and/or @darkk.
+      break;
+    }
+    // When the report ID is empty, do not bother with closing the report as
+    // it means we could not open it for some reason. An empty str instead
+    // indicates a bug where we could not serialize a JSON.
+    if (!settings_.no_collector && !ctx.report_id.empty() && !str.empty()) {
+      ErrContext err{};
+      if (!update_report(collector_base_url, ctx.report_id, str, info, &err)) {
```

Comment:
Is it the case that this new API doesn't write the report to disk anymore?

Reply:
Not in this header. My ideal scenario would be the one in which we drop all support for writing on disk from MK but I am not sure this is reasonable for others in the short terms. Hence, the more reasonable and backwards compatible plan is to write support for this functionality in the code that uses this header in MK.

Reply:
See 27a40efb3858bb4116ff9e27c1d6bcd5011b9579